File: android/app/src/main/java/com/synapse/app/core/qbank/QBankCollections.kt
package com.synapse.app.core.qbank

/** Session id -> the question ids that sitting served, in the order it served them. */
typealias SessionManifests = Map<String, List<String>>

/**
 * The questions a student can come back to: Flagged, Got wrong (incorrect), and
 * Omitted, plus the "Previous tests" list. Ported from `src/data/qbankCollections.ts`
 * and (for [bySession]) `src/data/attemptStats.ts`.
 *
 * One difference from the web port, forced by how [AttemptRecord] is written on
 * Android: [SessionViewModel][com.synapse.app.feature.qbank.SessionViewModel.finish]
 * writes one record per question in the sitting, whether it was answered or not —
 * an unanswered question's [AttemptRecord.correct] is `false`, never `null` (see
 * [QBankSession.finish]). Web's port instead skips a record with `correct === null`
 * to keep an unmarked/skipped station out of the wrong list. Android has no such
 * `null` to key off, so every function here gates on [AttemptRecord.selectedIndex]
 * being non-null instead — "was an option actually picked" rather than "was this
 * marked" — which is the same distinction web's `correct === null` check exists to
 * draw, just recovered from the field Android actually has.
 */
object QBankCollections {

    /** The surfaces whose records count as sitting a question from the bank. */
    private val QUESTION_SURFACES = setOf("qbank", "room")

    /**
     * How many sittings a manifest map keeps. Written only when a sitting begins
     * or a test is deleted — never per answer — so the bound is set high enough
     * that no real student reaches it.
     */
    const val MAX_STORED_SITTINGS = 600

    /**
     * The most recent verdict per question, among records that actually carry one
     * — see the class doc for why that is [AttemptRecord.selectedIndex] rather
     * than a null [AttemptRecord.correct].
     */
    fun latestVerdicts(records: List<AttemptRecord>): Map<String, Boolean> {
        data class Seen(val at: String, val correct: Boolean)
        val latest = HashMap<String, Seen>()
        for (record in records) {
            if (record.surface !in QUESTION_SURFACES || record.selectedIndex == null) continue
            val seen = latest[record.itemId]
            if (seen != null && seen.at >= record.at) continue
            latest[record.itemId] = Seen(record.at, record.correct == true)
        }
        return latest.mapValues { it.value.correct }
    }

    /** Questions whose latest verdict was wrong. Getting one right takes it out of this set. */
    fun incorrectIds(records: List<AttemptRecord>): Set<String> =
        latestVerdicts(records).filterValues { !it }.keys

    /**
     * Served by a sitting, and never actually answered anywhere since.
     *
     * A question the student has picked an option for at any point has engaged
     * with it, so it leaves this set whichever sitting the pick belongs to.
     */
    fun omittedIds(manifests: SessionManifests, records: List<AttemptRecord>): Set<String> {
        val answered = records
            .filter { it.surface in QUESTION_SURFACES && it.selectedIndex != null }
            .map { it.itemId }
            .toSet()
        return manifests.values.flatten().filterNot { it in answered }.toSet()
    }

    /** The manifests of sittings that are over — excludes [openSessionId], if any, from [manifests]. */
    fun finishedManifests(manifests: SessionManifests, openSessionId: String?): SessionManifests =
        if (openSessionId == null) manifests else manifests - openSessionId

    /** The published questions behind a set of ids, in pool order. */
    fun questionsById(pool: List<Question>, ids: Set<String>): List<Question> =
        pool.filter { it.id in ids }

    /**
     * The scope a set of questions implies — the topics they came from, not the
     * questions themselves, so "Test this scope" also offers material the student
     * has not seen.
     *
     * Simpler than web's port: [QBankScope.topicsFromQuestions] synthesizes
     * [topics] straight from the question set (there is no separate library tree
     * to cross-reference — see [QBankScope]'s doc comment), so every question's
     * topic is already one of [topics] by construction. A question naming a topic
     * outside [topics] (defensively handled, as web's port is) contributes
     * nothing.
     */
    fun scopeFromQuestions(questions: List<Question>, topics: List<TopicNode>): Set<String> {
        val topicIdByTitle = topics.associate { it.topic to it.id }
        return questions
            .mapNotNull { topicIdByTitle[it.topic.trim()] }
            .map(QBankScope::topicKey)
            .toSet()
    }

    /** Keep the newest [limit] sittings; returns [manifests] untouched when already within it. */
    fun pruneManifests(manifests: SessionManifests, limit: Int = MAX_STORED_SITTINGS): SessionManifests {
        if (manifests.size <= limit) return manifests
        return manifests.entries.toList().takeLast(limit).associate { it.key to it.value }
    }

    /** One finished sitting, reconstructed from the attempt records it produced. */
    data class SessionSummary(
        val sessionId: String,
        /** When the first answer in the sitting was recorded. */
        val startedAt: String,
        /** When the last one was. */
        val endedAt: String,
        val answered: Int,
        /** Answers actually picked — a crossed-out/never-touched station says nothing about correctness. */
        val marked: Int,
        val correct: Int,
        val accuracy: Double?,
        /** Distinct subjects covered, most-answered first. */
        val subjectIds: List<String>,
    )

    /** Group the attempt log into sittings, newest first. */
    fun bySession(records: List<AttemptRecord>): List<SessionSummary> {
        val groups = LinkedHashMap<String, MutableList<AttemptRecord>>()
        for (record in records) {
            if (record.sessionId.isBlank()) continue
            groups.getOrPut(record.sessionId) { mutableListOf() }.add(record)
        }
        return groups.map { (sessionId, group) ->
            val times = group.map { it.at }.sorted()
            val marked = group.filter { it.selectedIndex != null }
            val correct = marked.count { it.correct == true }
            val bySubject = group.groupingBy { it.subjectId }.eachCount()
            SessionSummary(
                sessionId = sessionId,
                startedAt = times.first(),
                endedAt = times.last(),
                answered = group.size,
                marked = marked.size,
                correct = correct,
                accuracy = if (marked.isNotEmpty()) correct.toDouble() / marked.size else null,
                subjectIds = bySubject.entries.sortedByDescending { it.value }.map { it.key },
            )
        }.sortedByDescending { it.startedAt }
    }

    /**
     * A synthetic, already-graded result built from the freshest attempt on
     * record for each of [questions] — lets a Collection (Flagged/Got
     * wrong/Omitted) be previewed through [com.synapse.app.feature.qbank.ResultsScreen]
     * without a live sitting. A question with no picked-and-recorded attempt shows
     * as unanswered, the same way [QBankSession.finish] represents one.
     */
    fun latestResultFor(questions: List<Question>, records: List<AttemptRecord>): QBankSessionResult =
        buildResult(questions) { question -> latestPick(question.id, records) }

    /** The result of one specific finished sitting, from its own attempt records only. */
    fun sessionResult(sessionId: String, questions: List<Question>, records: List<AttemptRecord>): QBankSessionResult {
        val own = records.filter { it.sessionId == sessionId }
        return buildResult(questions) { question -> latestPick(question.id, own) }
    }

    private fun latestPick(itemId: String, records: List<AttemptRecord>): AttemptRecord? =
        records
            .filter { it.itemId == itemId && it.surface in QUESTION_SURFACES && it.selectedIndex != null }
            .maxByOrNull { it.at }

    private fun buildResult(questions: List<Question>, pickFor: (Question) -> AttemptRecord?): QBankSessionResult {
        val perQuestion = questions.map { question ->
            val pickedLabel = pickFor(question)?.selectedIndex?.let { index -> question.options.getOrNull(index)?.label }
            QBankQuestionResult(
                questionId = question.id,
                pickedLabel = pickedLabel,
                correct = pickedLabel != null && question.isCorrect(pickedLabel),
            )
        }
        return QBankSessionResult(
            correct = perQuestion.count { it.correct },
            total = perQuestion.size,
            perQuestion = perQuestion,
        )
    }
}

File: src/data/qbankCollections.ts
import { attemptOrder, type AttemptRecord } from './attempts.ts'
import type { Question } from './qbank.ts'
import type { LibTopic } from './library.ts'
import { subtopicKey, topicKey, type Scope } from './qbankScope.ts'

/**
 * The questions a student can come back to.
 *
 * Three collections, and the evidence behind each one lives somewhere
 * different. Flags are their own stored list. Wrong answers are a query over
 * the attempt log. Omissions are neither: the log only receives a question once
 * its answer is checked, so a skipped question writes nothing at all, and the
 * only way to know one was skipped is to know what the sitting contained.
 * `SessionManifests` is that record.
 *
 * Getting a question right takes it out of the wrong list, and answering one
 * takes it out of the omitted list. Both are working sets a student can empty.
 * A permanent record of every mistake would only ever grow, which is the
 * opposite of something to revise from.
 */

/** Which questions each finished sitting contained, in the order they were sat. */
export type SessionManifests = Record<string, string[]>

/**
 * How many sittings the manifest map keeps.
 *
 * `usePersistentState` rewrites a whole document on every change, so this map
 * has to be bounded. It is written only when a sitting begins or a test is
 * deleted — never per answer — so the bound is set high enough that no real
 * student reaches it rather than low enough to keep the document small.
 */
export const MAX_STORED_SITTINGS = 600

/** The surfaces whose records count as sitting a question from the bank. */
const QUESTION_SURFACES: ReadonlySet<string> = new Set(['qbank', 'room'])

/**
 * The most recent marked verdict per question.
 *
 * Unmarked records are skipped rather than treated as wrong: a station is
 * ticked by the student against a checklist, so it is evidence of practice and
 * says nothing about correctness.
 */
export function latestVerdicts(records: AttemptRecord[]): Map<string, boolean> {
  const latest = new Map<string, { order: number; correct: boolean }>()
  for (const entry of records) {
    if (!QUESTION_SURFACES.has(entry.surface) || entry.correct === null) continue
    const order = attemptOrder(entry)
    const seen = latest.get(entry.itemId)
    if (seen && seen.order >= order) continue
    latest.set(entry.itemId, { order, correct: entry.correct })
  }
  const out = new Map<string, boolean>()
  latest.forEach((entry, itemId) => out.set(itemId, entry.correct))
  return out
}

export function incorrectIds(records: AttemptRecord[]): Set<string> {
  const out = new Set<string>()
  latestVerdicts(records).forEach((correct, itemId) => {
    if (!correct) out.add(itemId)
  })
  return out
}

/**
 * Served by a sitting, and never answered anywhere since.
 *
 * Order does not come into it. A question the student has attempted at any
 * point is one they have engaged with, so it leaves the list whichever sitting
 * the attempt belongs to — which also means this needs no timestamp for the
 * sitting itself, and a manifest is enough.
 */
export function omittedIds(manifests: SessionManifests, records: AttemptRecord[]): Set<string> {
  const answered = new Set<string>()
  for (const entry of records) {
    if (!QUESTION_SURFACES.has(entry.surface)) continue
    answered.add(entry.itemId)
  }
  const out = new Set<string>()
  for (const questionIds of Object.values(manifests)) {
    for (const id of questionIds) {
      if (!answered.has(id)) out.add(id)
    }
  }
  return out
}

/** Every question id the student has attempted on a question surface, ever. */
export function attemptedIds(records: AttemptRecord[]): Set<string> {
  const out = new Set<string>()
  for (const entry of records) {
    if (!QUESTION_SURFACES.has(entry.surface)) continue
    out.add(entry.itemId)
  }
  return out
}

/** The published questions behind a set of ids, in pool order. */
export function questionsById(pool: Question[], ids: Set<string>): Question[] {
  return pool.filter((question) => ids.has(question.id))
}

/**
 * The scope a set of questions implies — the topics they came from, not the
 * questions themselves.
 *
 * `questionsInScope` matches either a subtopic id carried by a library
 * reference or a topic title, so both kinds of key are emitted. A question
 * naming a topic outside the tree it is given contributes nothing; callers pass
 * the merged `chooserTopics` tree, where every topic the bank names exists.
 */
export function scopeFromQuestions(questions: Question[], libraryTopics: LibTopic[]): Scope {
  const byTitle = new Map<string, string>()
  for (const topic of libraryTopics) byTitle.set(topic.title.trim().toLowerCase(), topic.id)

  const scope: Scope = new Set()
  for (const question of questions) {
    for (const ref of question.libraryRefs) scope.add(subtopicKey(ref.id))
    const topicId = byTitle.get((question.topic ?? '').trim().toLowerCase())
    if (topicId) scope.add(topicKey(topicId))
  }
  return scope
}

/**
 * Keep the newest `limit` sittings.
 *
 * Insertion order is the chronological order: sittings are added as they are
 * started, and a session id is never an integer-like key, so both the object
 * and its JSON round-trip preserve it. Returns the input untouched when it is
 * already within the limit, so a write is only made when one is needed.
 */
export function pruneManifests(
  manifests: SessionManifests,
  limit = MAX_STORED_SITTINGS,
): SessionManifests {
  const keys = Object.keys(manifests)
  if (keys.length <= limit) return manifests
  const out: SessionManifests = {}
  for (const key of keys.slice(keys.length - limit)) out[key] = manifests[key]
  return out
}
